File: swift/Sources/MoneyHero/Dashboard/DashboardState.swift
import Combine
import Foundation

@MainActor
final class DashboardState: ObservableObject {
    @Published var snapshot: MarketRefreshSnapshot
    @Published var isAppOpenLoading: Bool
    @Published var isInitialLoading: Bool
    @Published var isRefreshing: Bool
    @Published var refreshErrorMessage: String?

    let currencySettings: CurrencyWidgetSetting
    let marketTickerSettings: MarketTickerSetting

    private let coordinator: MarketRefreshCoordinator?
    private let holdings: [Holding]
    private var snapshotsTask: Task<Void, Never>?
    private var awaitingRefreshProgress: Bool
    private var hasStartedAppOpenRefresh: Bool

    init(
        coordinator: MarketRefreshCoordinator?,
        holdings: [Holding] = [],
        currencySettings: CurrencyWidgetSetting = .default,
        marketTickerSettings: MarketTickerSetting = .default,
        startupErrorMessage: String? = nil
    ) {
        self.coordinator = coordinator
        self.holdings = holdings
        self.currencySettings = currencySettings
        self.marketTickerSettings = marketTickerSettings
        self.refreshErrorMessage = startupErrorMessage
        self.snapshot = .empty
        self.isAppOpenLoading = true
        self.isInitialLoading = false
        self.isRefreshing = false
        self.awaitingRefreshProgress = false
        self.hasStartedAppOpenRefresh = false

        observeSnapshots()
    }

    deinit {
        snapshotsTask?.cancel()
    }

    static func live() -> DashboardState {
        do {
            let cacheStore = try LocalJSONCacheStore()
            return DashboardState(coordinator: MarketRefreshCoordinator(cacheStore: cacheStore))
        } catch {
            return DashboardState(
                coordinator: nil,
                startupErrorMessage: "Cache unavailable: \(String(describing: error))"
            )
        }
    }

    var hasHoldings: Bool {
        holdings.contains { !$0.isHidden }
    }

    func refreshOnAppOpen() async {
        guard !hasStartedAppOpenRefresh else {
            return
        }

        hasStartedAppOpenRefresh = true
        isAppOpenLoading = true
        guard shouldStartRefresh else {
            isAppOpenLoading = false
            return
        }

        isInitialLoading = !hasAnyMarketData
        await refresh { coordinator in
            await coordinator.refreshOnAppOpen(
                holdings: holdings,
                currencySettings: currencySettings,
                marketTickerSettings: marketTickerSettings
            )
        }
    }

    func refreshOnPullToRefresh() async {
        guard shouldStartRefresh else {
            return
        }

        await refresh { coordinator in
            await coordinator.refreshOnPullToRefresh(
                holdings: holdings,
                currencySettings: currencySettings,
                marketTickerSettings: marketTickerSettings
            )
        }
    }

    private var shouldStartRefresh: Bool {
        !isRefreshing && coordinator != nil
    }

    private var hasAnyMarketData: Bool {
        !snapshot.quotesBySymbol.isEmpty || !snapshot.historiesBySymbol.isEmpty || snapshot.lastUpdatedAtMs != nil
    }

    private func refresh(_ operation: (MarketRefreshCoordinator) async -> Void) async {
        guard let coordinator else {
            refreshErrorMessage = refreshErrorMessage ?? "Market data is unavailable."
            isInitialLoading = false
            isRefreshing = false
            isAppOpenLoading = false
            return
        }

        isRefreshing = true
        awaitingRefreshProgress = true
        refreshErrorMessage = nil
        await operation(coordinator)

        snapshot = await coordinator.snapshot()
        awaitingRefreshProgress = false
        applyProgress(snapshot.progress)
    }

    private func observeSnapshots() {
        guard let coordinator else {
            return
        }

        snapshotsTask = Task { [weak self, coordinator] in
            for await snapshot in coordinator.snapshots() {
                guard let self else {
                    return
                }

                self.snapshot = snapshot
                self.refreshErrorMessage = Self.errorMessage(from: snapshot)
                self.applyProgress(snapshot.progress)
            }
        }
    }

    private func applyProgress(_ progress: MarketFetchQueueProgress) {
        if progress.isActive {
            awaitingRefreshProgress = false
            isRefreshing = true
            return
        }

        if progress.totalCount > 0 {
            awaitingRefreshProgress = false
            isRefreshing = false
            isInitialLoading = false
            isAppOpenLoading = false
            return
        }

        guard !awaitingRefreshProgress else {
            return
        }

        isRefreshing = false
        isInitialLoading = false
        isAppOpenLoading = false
    }

    private static func errorMessage(from snapshot: MarketRefreshSnapshot) -> String? {
        let quoteError = snapshot.quotesBySymbol.values.first { $0.metadata.status == .error }?.metadata.error
        let historyError = snapshot.historiesBySymbol.values.first { $0.metadata.status == .error }?.metadata.error
        return snapshot.refreshErrorMessage ?? quoteError ?? historyError
    }
}

File: flutter/lib/dashboard/dashboard_state.dart
import "dart:async";

import "package:flutter/foundation.dart";

import "../cache/local_json_cache_store.dart";
import "../domain/holdings_models.dart";
import "../market_data/market_fetch_queue.dart";
import "../market_data/market_refresh_coordinator.dart";
import "../market_data/market_refresh_snapshot.dart";

class DashboardState extends ChangeNotifier {
  DashboardState({
    required MarketRefreshCoordinator? coordinator,
    this.holdings = const <Holding>[],
    this.currencySettings = CurrencyWidgetSetting.defaults,
    this.marketTickerSettings = MarketTickerSetting.defaults,
    this.refreshErrorMessage,
  }) : _coordinator = coordinator {
    _observeSnapshots();
  }

  factory DashboardState.live() {
    try {
      return DashboardState(coordinator: MarketRefreshCoordinator(cacheStore: LocalJsonCacheStore()));
    } catch (error) {
      return DashboardState(coordinator: null, refreshErrorMessage: "Cache unavailable: $error");
    }
  }

  final MarketRefreshCoordinator? _coordinator;
  final List<Holding> holdings;
  final CurrencyWidgetSetting currencySettings;
  final MarketTickerSetting marketTickerSettings;
  MarketRefreshSnapshot snapshot = MarketRefreshSnapshot.empty;
  bool isAppOpenLoading = true;
  bool isInitialLoading = false;
  bool isRefreshing = false;
  String? refreshErrorMessage;
  bool _hasStartedAppOpenRefresh = false;
  StreamSubscription<MarketRefreshSnapshot>? _snapshotSubscription;

  bool get hasHoldings => holdings.any((Holding h) => !h.isHidden);

  Future<void> refreshOnAppOpen() async {
    if (_hasStartedAppOpenRefresh) {
      return;
    }
    _hasStartedAppOpenRefresh = true;
    isAppOpenLoading = true;
    notifyListeners();
    await _refresh((MarketRefreshCoordinator coordinator) {
      return coordinator.refreshOnAppOpen(
        holdings: holdings,
        currencySettings: currencySettings,
        marketTickerSettings: marketTickerSettings,
      );
    });
  }

  Future<void> refreshOnPullToRefresh() {
    return _refresh((MarketRefreshCoordinator coordinator) {
      return coordinator.refreshOnPullToRefresh(
        holdings: holdings,
        currencySettings: currencySettings,
        marketTickerSettings: marketTickerSettings,
      );
    });
  }

  Future<void> _refresh(Future<void> Function(MarketRefreshCoordinator coordinator) operation) async {
    if (isRefreshing) {
      return;
    }
    if (_coordinator == null) {
      refreshErrorMessage ??= "Market data is unavailable.";
      isInitialLoading = false;
      isRefreshing = false;
      isAppOpenLoading = false;
      notifyListeners();
      return;
    }
    isRefreshing = true;
    isInitialLoading = snapshot.quotesBySymbol.isEmpty && snapshot.historiesBySymbol.isEmpty;
    refreshErrorMessage = null;
    notifyListeners();
    await operation(_coordinator!);
  }

  void _observeSnapshots() {
    if (_coordinator == null) {
      return;
    }
    _snapshotSubscription = _coordinator!.snapshots().listen((MarketRefreshSnapshot value) {
      snapshot = value;
      refreshErrorMessage = value.refreshErrorMessage ??
          value.quotesBySymbol.values.where((e) => e.metadata.error != null).map((e) => e.metadata.error).cast<String?>().firstWhere((_) => true, orElse: () => null);
      _applyProgress(value.progress);
      notifyListeners();
    });
  }

  void _applyProgress(MarketFetchQueueProgress progress) {
    if (progress.isActive) {
      isRefreshing = true;
      return;
    }
    isRefreshing = false;
    isInitialLoading = false;
    isAppOpenLoading = false;
  }

  @override
  void dispose() {
    _snapshotSubscription?.cancel();
    super.dispose();
  }
}
